reactive/kotlinx-coroutines-rx2/test/SchedulerTest.kt

```diff
@@ -31,4 +38,252 @@ class SchedulerTest : TestBase() {
+        val scheduler = (currentDispatcher() as CoroutineDispatcher).asScheduler()
+        suspendCancellableCoroutine<Unit> {
+            scheduler.scheduleDirect {
+                val t1 = Thread.currentThread()
+                assertSame(t1, mainThread)
+                expect(2)
+                it.resume(Unit)
+            }
+        }
+        yield()
+        finish(3)
+    }
+
+    @Test
+    fun testAsSchedulerWithDelay(): Unit = runTest {
+        expect(1)
+        val mainThread = Thread.currentThread()
+        val scheduler = (currentDispatcher() as CoroutineDispatcher).asScheduler()
+        val delayMillis = 300L
+        suspendCancellableCoroutine<Unit> {
+            scheduler.scheduleDirect({
+                val t1 = Thread.currentThread()
+                assertSame(t1, mainThread)
+                expect(2)
+                it.resume(Unit)
+            }, delayMillis, TimeUnit.MILLISECONDS)
+        }
+        finish(3)
+    }
+
+    @Test
+    fun testAsSchedulerWithZeroDelay(): Unit = runTest {
+        expect(1)
+        val mainThread = Thread.currentThread()
+        val scheduler = (currentDispatcher() as CoroutineDispatcher).asScheduler()
+        suspendCancellableCoroutine<Unit> {
+            scheduler.scheduleDirect({
+                val t1 = Thread.currentThread()
+                assertSame(t1, mainThread)
+                expect(2)
+                it.resume(Unit)
+            }, 0, TimeUnit.MILLISECONDS)
+        }
+
+        scheduler.shutdown()
+        finish(3)
+    }
+
+    @Test
+    fun testAsSchedulerWithNegativeDelay(): Unit = runTest {
+        expect(1)
+        val mainThread = Thread.currentThread()
+        val scheduler = (currentDispatcher() as CoroutineDispatcher).asScheduler()
+        suspendCancellableCoroutine<Unit> {
+            scheduler.scheduleDirect({
+                val t1 = Thread.currentThread()
+                assertSame(t1, mainThread)
+                expect(2)
+                it.resume(Unit)
+            }, -1, TimeUnit.MILLISECONDS)
+        }
+        yield()
+        finish(3)
+    }
+
+    @Test
+    fun testDisposeDuringDelay(): Unit = runTest {
+        expect(1)
+        val scheduler = (currentDispatcher() as CoroutineDispatcher).asScheduler()
+        val delayMillis = 300L
+        val disposable = scheduler.scheduleDirect({
+            expectUnreached()
+        }, delayMillis, TimeUnit.MILLISECONDS)
+        delay(100)
+        expect(2)
+        disposable.dispose()
+        delay(300)
+        yield()
+        finish(3)
+    }
+
+    @Test
+    fun testImmediateDispose(): Unit = runTest {
+        expect(1)
+        val scheduler = (currentDispatcher() as CoroutineDispatcher).asScheduler()
+        val disposable = scheduler.scheduleDirect {
+            expectUnreached()
+        }
+        disposable.dispose()
+        yield()
+        finish(2)
+    }
+
+    @Test
+    fun testAsSchedulerWorksWithSchedulerCoroutineDispatcher(): Unit = runTest {
```

Comment:
All these tests for `scheduleDirect` should also be duplicated for tasks scheduled via workers.